### PR TITLE
HDDS-10411. Support incremental ChunkBuffer checksum calculation

### PR DESCRIPTION
**Pull request by smengcl**
### Problem Statement

Currently by default, each 4 MB block chunk is further divided into 16 KB chunks (down from 1 MB, changed in HDDS-10465) for checksum calculation.

The problem is, even with the smaller checksum chunk, clients still calculate the checksum for the whole 4 MB block chunk **from the beginning** every single time:

https://github.com/apache/ozone/blob/e57370124a36315d2be5791753912901f836ccd8/hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/common/Checksum.java#L171-L177

This PR aims to implement a checksum cache to reduce the CPU time spent in critical section for checksum calculation, in hope of greatly improving client hsync throughput when checksum is enabled.

### TODOs

- [x] Add client config key to enable client checksum cache?
- [ ] Thoroughly test all code paths using `Checksum`
- [x] Verify boundary calc in edge cases

### Future Work

There are much more improvements that can be done on optimizing checksum, such as:
1. Even finer-grained incremental checksum calculation.
  - For CRC32/CRC32C, the checksum can be updated on a byte-by-byte basis (rather having to calculate the entire 16 KB)
  - For SHA256 and MD5, the checksum can be updated every 64 bytes (512 bits).
2. Transfer only the unacknowledged checksums to the datanode. Requires proto change.

But those are beyond the scope of this jira and would require major refactoring.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-10411

## How was this patch tested?

- [x] New UT cases to be added

### CODE REVIEW

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/common/ChecksumCache.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        i++;
+        continue;
+      }
+
+      // i can either point to:
+      // 1. the last element in the list -- in which case the checksum needs to be updated
+      // 2. one after the last element   -- in which case a new checksum needs to be added
+      assert i == checksums.size() - 1 || i == checksums.size();
+
+      // TODO: Furthermore for CRC32/CRC32C, it can be even more efficient by updating the last checksum byte-by-byte.
```
Comment (jojochuang) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/common/ChecksumCache.java:95]: The checksum functions don't allow updating checksum, so this is going to require a refactoring.

**Review by jojochuang — COMMENTED** [review]
The patch makes sense to me overall. I think you also need to switch BlockOutputStream to use ChecksumCache.
Also, is it thread safe?

**Comment by jojochuang** [comment]
Looks like all tests are passing now except for an unrelated failure.

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/OzoneClientConfig.java [hunk 2]
```diff
@@ -178,6 +178,15 @@ public enum ChecksumCombineMode {
       tags = { ConfigTag.CLIENT, ConfigTag.CRYPTO_COMPLIANCE })
   private int bytesPerChecksum = 16 * 1024;
 
+  // Client-side block chunk checksum cache config
+  // TODO: Re-evaluate if this need to be exposed to end-users at all, e.g. to avoid confusion.
+  @Config(key = "chunk.checksum.cache.enabled",
```
Comment (jojochuang) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/OzoneClientConfig.java:183]: IMO this should be the default behavior if incremental chunk list option is enabled.
Ok to simply drop this configuration property

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/common/ChecksumCache.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+ */
+public class ChecksumCache {
+  public static final Logger LOG = LoggerFactory.getLogger(ChecksumCache.class);
+
+  private final int bytesPerChecksum;
+  private final List<ByteString> checksums;
+  // Chunk length last time the checksum is computed
+  private int prevChunkLength;
+  // This only serves as a hint for array list initial allocation. The array list will still grow as needed.
+  private static final int BLOCK_CHUNK_SIZE = 4 * 1024 * 1024; // 4 MB
```
[the remark this answers is not in the document]
Comment (smengcl) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/ozone/common/ChecksumCache.java:46]: Yes. Chunk size is 4 MB by default.

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/OzoneClientConfig.java (hunk 2 above)
Comment (smengcl) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/OzoneClientConfig.java:183]: Agreed. I added this to facilitate testing (easier switching it on/off for comparison).

**Comment by jojochuang** [comment]
Before:
Screenshot 2024-11-14 at 2 52 47 PM

After: (with ozone.client.bytes.per.checksum = 16KB)
Screenshot 2024-11-14 at 2 28 07 PM

So substantial improvement in CPU time.

YCSB benchmark on my test cluster shows throughput improved from ~800 ops per second to ~1500 ops per second.

**Review by jojochuang — APPROVED** [review]
+1
This is quite substantial improvement over before. Now the bottomneck appear to shift to the DataNode so let's get this in and then work on DataNode optimization.

**Comment by smengcl** [comment]
Thanks @jojochuang . I will do a final cleanup on this jira and then I will merge it.
